_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_hw.cpp

```diff
@@ -2465,6 +2468,9 @@ mfxStatus ImplementationAvc::AsyncRoutine(mfxBitstream * bs)
         DdiTaskIter task = FindFrameToStartEncode(m_video, m_lookaheadFinished.begin(), m_lookaheadFinished.end());
         assert(task != m_lookaheadFinished.end());
 
+        task->m_decodeTimeStamp = m_decodeTimeStamp.front();
```

Comment:
Maybe I'm missing something (haven't been looking to this code for years). But with this implementation few first encoded frames will have DecodeTimeStamp = MFX_TIMESTAMP_UNKNOWN. 

Below is example for B-pyramid for 3 B-frames: I b B b P b B b P

Number of EncodeFrameAync call |  1 |   2 |   3 |   4 |   5 |   6 |   7 |   8 |   9
Input frame with PTS                      | I0 | b1 | B2 | b3 | P4 | b5 | B6 | b7 | P8
Encoded frame                               |     |      |       | I0 |  P4 | B2 | b1 | b3 | P8
DTS assigned to encoded frame    |      |     |        | U |   U |   0 |   1 |   2 |   3

It's my main concern. There are few other (mostly to look code nicer and avoid multiple implementations of same logic). But let's deal with this one first.

Reply:
Yes, I set the few first encoded frames with unkown DTS. From frame's  PTS information, there is not a good way to get the correct DTS for the first few frames. And it is not necessary that all frames should have DTS. It is better to let few first frame's DTS as unknown DTS, so that decoder side can handle it itself. Do you have better idea here?

Reply:
Actually I don't understand why we need this intentional delay filled with MFX_TIMESTAMP_UNKNOWN.  Why in the example above coded B2 should have DST = 0? E.g. in previous logic I0 would have DTS 0 and P4 would have DTS 1.

Reply:
Your understanding is not the case. If I0 have DTS 0 and P4 have DTS 1, then b1 would have DTS 4, but b1 have PTS 1. It is not correct that b1's DTS is bigger than PTS. 
In fact, the intentional delay is to make every frame's DTS is smaller than PTS. In this B pyramid, we need delay two frames unit so that the b1 have DTS 1 and PTS 1.

Reply:
I encoded few streams with different GOP structures, carefully wrote all numbers and see that you are right (there are no problems with intentional delay as I originally thought). E.g. below is table for 3B-frame B-pyramid:

xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx  |xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx
------------------------------  | -- | -- | -- | -- | -- | -- | -- | -- | --
Number of EncodeFrameAync call  |  1 |  2 |  3 |  4 |  5 |  6 |  7 |  8 |  9
Input frame (with display order)| I0 | b1 | B2 | b3 | P4 | b5 | B6 | b7 | P8
PTS (linear)                    | 3K | 6K | 9K | 12K| 15K| 18K| 21K| 24K| 27K 
PTS (jump)                      | 3K | 6K | 90K| 93K| 96K| 99K|102K|105K|108K 
Encoded frame                   |    |    |    | I0 | P4 | B2 | b1 | b3 | P8
cpb_removal_delay               |    |    |    |  0 |  2 |  4 |  6 |  8 | 10 
dpb_output_delay                |    |    |    |  4 | 10 |  4 |  0 |  2 | 10 
dpb output time (HRD)               |    |    |    |  4 | 12 |  8 |  6 | 10 | 20 
DTS (linear) - current          |    |    |    | -3K|  0 | 3K | 6K | 9K | 12K
DTS (jump) - current            |    |    |    | -3K| 81K| 84K| 6K | 90K| 93K
DTS (linear) - Hongbo           |    |    |    | Unk| Unk| 3K | 6K | 9K | 12K
DTS (jump) - Hongbo             |    |    |    | Unk| Unk| 3K | 6K | 90K| 93K

It also shows that "DTS = reordered PTS" fixes "jittering" of DTS in case of PTS jumps.
So concern with algorithm itself is resolved. Will share more concerns about implementation and customer-side behavior in next comments.